src/main/java/org/commonjava/util/partyline/impl/infinispan/model/InfinispanJF.java

```diff
@@ -0,0 +1,615 @@
+                currBlock = block;
+
+                flush();
+
+                currBlock.writeToBuffer( (byte) b );
+            }
+        }
+
+        public void write( ByteBuffer buf ) throws IOException
+        {
+            while ( buf.hasRemaining() )
+            {
+                byte b = buf.get();
+                write( (int) b );
+            }
+        }
+
+        @Override
+        public void flush() throws IOException
+        {
+            synchronized ( InfinispanJF.this )
+            {
+                if ( closed )
+                {
+                    throw new IOException( "Cannot write to closed stream!" );
+                }
+            }
+
+            doFlush( false );
+        }
+
+        @Override
+        public void close() throws IOException
+        {
+            Logger logger = LoggerFactory.getLogger( getClass() );
+            logger.trace( "OUT ({}):: close() called", originalThreadName );
+
+            if ( closed )
+            {
+                logger.trace( "OUT ({}):: already closed", originalThreadName );
+                return;
+            }
+            doFlush( true );
+            super.close();
+
+            InfinispanJF.this.close();
+        }
+
+        public void doFlush( boolean eof ) throws IOException
+        {
+            if ( eof )
+            {
+                filesystem.pushNextBlock( prevBlock, currBlock, metadata );
+            }
+            else
+            {
+                filesystem.updateBlock( prevBlock );
+            }
+
+            synchronized ( InfinispanJF.this )
+            {
+                InfinispanJF.this.notifyAll();
+            }
+
+            if ( callbacks != null )
+            {
+                callbacks.flushed();
+            }
+        }
+
+        boolean isClosed()
+        {
+            return closed;
+        }
+    }
+
+    /**
+     * {@link InputStream} associated with a particular {@link JoinableFile} instance. This stream reads content that the output stream has
+     * already flushed to disk, and waits for new content to become available (or for the output stream to close). This allows multiple readers
+     * when content is still being written to disk.
+     */
+    private final class JoinInputStream
+                    extends InputStream
+    {
+        private static final long MAX_BUFFER_SIZE = 5 * 1024 * 1024; // 5Mb.
+
+        private long read = 0;
+
+        private FileBlock block;
+
+        private boolean closed = false;
+
+        private final int jointIdx;
+
+        private final String originalThreadName;
+
+        private final long ctorTime;
+
+        /**
+         * Map the content already written to disk for reading. If the flushed count exceeds MAX_BUFFER_SIZE, use the max instead.
+         */
+        JoinInputStream( int jointIdx ) throws IOException
+        {
+            this.jointIdx = jointIdx;
+            block = metadata.getFirstBlock();
```

Comment:
This will have to change to InfinispanJFS.getFirstBlock(metadata) I think, or something like it.